pre_commit/commands/run.py

```diff
@@ -261,14 +275,17 @@ def _run_hooks(
         _all_filenames(args), config['files'], config['exclude'],
     )
     retval = 0
+    prior_diff: Optional[Tuple[int, bytes, Optional[bytes]]] = None
     for hook in hooks:
-        retval |= _run_single_hook(
-            classifier, hook, skips, cols,
+        current_retval, current_diff = _run_single_hook(
+            classifier, hook, skips, cols, prior_diff,
             verbose=args.verbose, use_color=args.color,
         )
+        retval |= current_retval
+        prior_diff = current_diff
         if retval and config['fail_fast']:
             break
-    if retval and args.show_diff_on_failure and git.has_diff():
+    if retval and args.show_diff_on_failure and _has_diff(prior_diff):
```

Comment:
with the suggestion above, this can become `and prior_diff` -- though ~generally this isn't in the critical path for committing (since `--show-diff-on-failure` is not accessible there) so maybe this change isn't necessary

also if this is the last caller of `git.has_diff()` can you delete that code?

Reply:
Not the last user - https://github.com/pre-commit/pre-commit/blob/master/pre_commit/commands/try_repo.py#L28